src/caffe/layers/data_layer.cpp

```diff
@@ -83,13 +83,25 @@ void DataLayer<Dtype>::InternalThreadEntry() {
   CPUTimer timer;
   CHECK(this->prefetch_data_.count());
   CHECK(this->transformed_data_.count());
+
+  // Reshape on single input batches for inputs of varying dimension.
+  const int batch_size = this->layer_param_.data_param().batch_size();
+  const int crop_size = this->layer_param_.transform_param().crop_size();
+  if (batch_size == 1 && crop_size == 0) {
+    Datum datum;
+    datum.ParseFromString(cursor_->value());
+    this->prefetch_data_.Reshape(1, datum.channels(),
+        datum.height(), datum.width());
+    this->transformed_data_.Reshape(1, datum.channels(),
```

Comment:
This is not a comment about this PR, but the code that follows is pretty confusing wrt `transformed_data_`: `prefetch_data_` becomes `top_data`, which then becomes `transformed_data_`... somehow they're all the same!

We should probably at least rename `top_data` and maybe eliminate `transformed_data_`, (but not in this PR).